### PR TITLE
New Blob Encoding, Re: Proto's Specs on Encoding Blobs More Efficiently

### PR DESCRIPTION
**Pull request by angel-ding-cb**
**Description**

This is an implementation of Proto's [specs] (https://github.com/ethereum-optimism/optimism/pull/8657) on the new blob encoding method for 4844. 

Instead of ignoring 1 byte of each field element, we now combine 4 of those remaining bytes together and encode 3 bytes of data into it. 

**Tests**

Wrote tests for small and large data encoding and decoding and invalid blobs.

**Metadata**

- More detailed encoding scheme #[[Link to Proto's Explanation](https://github.com/ethereum-optimism/optimism/pull/8657/files)]

### CODE REVIEW

File: op-service/eth/blob.go [hunk 1]
```diff
@@ -69,60 +71,181 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+		decodedData[1] = b[FieldSize*i+32] & 0b0011_1111
+		decodedData[2] = b[FieldSize*i+64] & 0b0011_1111
+
+		// Extract the first 2 bits of all remainingData bytes from the first byte of the fourth field element
+		decodedData[0] |= (b[FieldSize*i+96] & 0b0011_0000) << 2
+		decodedData[1] |= (b[FieldSize*i+96] & 0b0000_1100) << 4
+		decodedData[2] |= (b[FieldSize*i+96] & 0b0000_0011) << 6
+
+		if i == 1023 {
+			fmt.Println("decodedData:", decodedData)
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go:240]: looks like you have some print statements to remove :)

File: op-service/eth/blob_test.go [hunk 2]
```diff
@@ -54,12 +54,12 @@ func TestInvalidBlobDecoding(t *testing.T) {
 	if err := b.FromData(data); err != nil {
 		t.Fatalf("failed to encode bytes: %v", err)
 	}
-	b[32] = 0x80 // field elements should never have their highest order bit set
+	b[1] = 0x01 // wrong version of encoding
```
Comment (roberto-bayardo) [inline op-service/eth/blob_test.go:57]: we should test both of these things (high order bit should never be set, as well as wrong encoding)

File: op-service/eth/blob.go [hunk 3]
```diff
@@ -69,60 +71,181 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+
+	// encode the length as big-endian uint24 into [2:5] bytes of the first field element
+	if len(data) < 1<<24 {
+		// Zero out any trailing data in the buffer if any
+		b[2] = byte((len(data) >> 16) & 0xFF) // Most significant byte
+		b[3] = byte((len(data) >> 8) & 0xFF)
+		b[4] = byte(len(data) & 0xFF) // Least significant byte
+		fmt.Println("FromData:", b[1:5])
+	} else {
+		// Handle the error for length_rollup_data being too large
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go:96]: comment doesn't add any additional info, suggest deleting.

File: op-service/eth/blob.go [hunk 4]
```diff
@@ -69,60 +71,181 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+	// encode the first 27 bytes of data into remaining bytes of first field element
 	offset := copy(b[5:32], data)
-	// encode (up to) 31 bytes of remaining input data at a time into the subsequent field element
-	for i := 1; i < 4096; i++ {
+
+	// for loop to encode the next 31 bytes of data into [1:] of the next three field elements
+	for i := 1; i < 4; i++ {
 		offset += copy(b[i*32+1:i*32+32], data[offset:])
+		if offset == len(data) {
+			return nil
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go:108]: Do the existing test cases that exercise this edge case?

File: op-service/eth/blob.go [hunk 5]
```diff
@@ -69,60 +71,181 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
...
+		}
+	}
+	// copy the last 6 bits of remainingData[0] into the first byte of the first field element
+	b[0] = remainingData[0] & 0b0011_1111
+	// copy the last 6 bits of remainingData[1] into the first byte of the second field element
+	b[32] = remainingData[1] & 0b0011_1111
+	// copy the last 6 bits of remainingData[2] into the first byte of the third field element
+	b[64] = remainingData[2] & 0b0011_1111
+	// copy the first 2 bits of all remainingData bytes into the first byte of the fourth field element
+	b[96] = ((remainingData[0] & 0b1100_0000) >> 2) | ((remainingData[1] & 0b1100_0000) >> 4) | ((remainingData[2] & 0b1100_0000) >> 6)
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go:127]: how about factoring these out into a helper function that you can use here and on L153?

File: op-service/eth/blob_test.go (hunk 2 above)
Comment (angel-ding-cb) [reply op-service/eth/blob_test.go:57]: I thought the higher order bit is now used to store the extra 3 bytes of data. For example, b[0], b[32], b[64] and b[96] are now used to store the next 3 bytes of data, after 27+31*3 bytes of data are encoded. Maybe my understanding is wrong here, could you help me confirm that?

Comment (roberto-bayardo) [reply op-service/eth/blob_test.go:57]: Bit not byte :)   

Proto's  spec:

> Instead, we can also choose to only use 254 bits per element:

So bits 254 & 255 should always be unset

Comment (angel-ding-cb) [reply op-service/eth/blob_test.go:57]: Ahh got it, sorry that I misread your comment!

File: op-service/eth/blob_test.go [hunk 6]
```diff
@@ -32,10 +32,56 @@ func TestBlobEncodeDecode(t *testing.T) {
...
+	}
+	if string(decoded) != string(data) {
+		t.Errorf("decoded blob != small blob input")
+	}
+}
+
 func TestBigBlobEncoding(t *testing.T) {
-	bigData := Data(make([]byte, MaxBlobDataSize))
-	bigData[MaxBlobDataSize-1] = 0xFF
+	bigData := Data(make([]byte, MaxBlobDataSize-3))
```
Comment (roberto-bayardo) [inline op-service/eth/blob_test.go:81]: why -3 here?

File: op-service/eth/blob.go [hunk 7]
```diff
@@ -14,7 +13,10 @@ import (
 
 const (
 	BlobSize        = 4096 * 32
-	MaxBlobDataSize = 4096*31 - 4
+	MaxBlobDataSize = (4*31+3)*1024 - 4
+	EncodingVersion = 0
+	FieldSize       = 4 * 32   // size of a field composed of 4 field elements in bytes
+	FieldCapacity   = 31*4 + 3 // maximum data bytes that can be encoded in a field
```
Comment (roberto-bayardo) [inline op-service/eth/blob.go:19]: 127 bytes ...I think you  mean # of bytes that can be encoded in **4** field elements?

File: op-service/eth/blob_test.go (hunk 6 above)
Comment (angel-ding-cb) [reply op-service/eth/blob_test.go:81]: I forgot to delete some manual debugging :)
